### PR TITLE
Remove bad implementation of functional grad

### LINKED ISSUE #13008: Computing conv grads via nn.grad raises ValueError when using dilation>1
## 🐛 Bug

The conv grads computed by https://github.com/pytorch/pytorch/blob/master/torch/nn/grad.py rely on the function ```_grad_input_padding()``` (from the same file). ```_grad_input_padding()``` raises a ValueError when ```dilation```>1 because  it does not account for ```dilation``` when computing its expected input size.

## To Reproduce

The following snippet appears to correctly create a grad with respect to the input  via ```torch.autograd.grad``` with dilation=2 (stored in ```grad_input```). We should be able to get the same grad by using ```torch.nn.grad.conv2d_input()```, but attempting to do so leads to a ValueError raised by ```_grad_input_padding()``` when dilation>1:

```python
import torch
import torch.nn.functional as F
input = torch.randn(1,1,5,5, requires_grad=True)
weight = torch.randn(1,1,3,3, requires_grad=True)
output = F.conv2d(input, weight, dilation=2)
grad_output = torch.randn(output.shape)
grad_input = torch.autograd.grad(output, input, grad_output)
torch.nn.grad.conv2d_input(input.shape, weight, grad_output, dilation=2)
```

## Error message: 

>File "/home/.../anaconda3/lib/python3.7/site-packages/torch/nn/grad.py", line 29, **in _grad_input_padding**
    grad_output.size()[2:]))

>ValueError: requested an input grad size of [5, 5], but valid sizes range from [3, 3] to [3, 3] (for a grad_output of torch.Size([1, 1]))

[As we would expect with a 5x5 input, 3x3 kernel, and dilation=2](https://ezyang.github.io/convolution-visualizer/index.html), the output size is [1,1]. According to ```_grad_input_padding()```, the only possible input size was [3,3]. I believe the fix is to make ```_grad_input_padding()``` account for ```dilation```. In this case, accounting for dilation would allow this function to know that  [5,5] is the only acceptable input size.

 - PyTorch 0.4.1 installed via conda
 - Ubuntu 18
 - Python 3.7

### PR DESCRIPTION
**Pull request by albanD**
Even though they compute what they claim, these functions are not representative of the actual backward pass that is used and so they are very confusing for the users.
This is a BC change but these functions can be copy/pasted in the user code if they need them.

Fix #13008 #16012

### CODE REVIEW

**Comment by kostmo** [comment]
## CircleCI build failures summary
As of commit 623a39f:
* **1/1** failures introduced in this PR
* **0/1** recognized as flaky



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=623a39ff330caf95ec54a861a23c1570ff6856a6):

| Job | Step | Log excerpt |
| --- | --- | --- |


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

**Comment by zou3519** [comment]
Should we do a deprecation cycle first or are these so egregious that we should remove them upfront like this?

**Comment by albanD** [comment]
@soumith recommended straight deletion.
The main arguments are that they are very scarcely used, not doing what they claim (they don't use backward kernels for cudnn in particular) and very easy to replace (just copy paste from github).

**Review by zou3519 — APPROVED** [review]
I'm happy to see these go as long as we're fine with just deleting these

File: torch/nn/grad.py [hunk 1]
```diff
@@ -1,307 +1,16 @@
 """Gradient interface"""
```
Comment (zou3519) [inline torch/nn/grad.py]: Maybe file an issue reminding ourselves to fully delete the file in 1.5, and link it from here?

Comment (albanD) [reply torch/nn/grad.py]: Well, having a functional interface for gradients would be nice I think. In particular for people implementing custom `Function`s that are similar to the original one.
But it should point to the functions that actually run during the regular backward.

File: torch/nn/grad.py [hunk 2]
```diff
@@ -1,307 +1,16 @@
...
-                4, 0, weight_size[4])
+def _get_not_implemented_function(func_name):
+    def tmp(*args, **kwargs):
+        err_message = "{} has been removed as it is not representative of the actual backward.\n"
+        err_message += "If you need the old implementation of this function, you can copy/paste "
+        err_message += "it from https://github.com/pytorch/pytorch/blob/v1.3.1/torch/nn/grad.py"
+        raise RuntimeError(err_message.format(func_name))
+    return tmp
+
+conv1d_input = _get_not_implemented_function("conv1d_input")
```
Comment (zou3519) [inline torch/nn/grad.py]: Might be good to add a test that calls conv1d_input and asserts that a RuntimeError is raised

Comment (albanD) [reply torch/nn/grad.py]: Done

**Review by soumith — APPROVED** [review]

**Comment by albanD** [comment]
The `_grad_input_padding` function is used by crypten at least.
These might be used quite a bit in the wild.
I think a better approach would be to replace these with the actual backward implementations.
cc @soumith

**Pull request closed without merging**